### ISSUE TITLE
Clarification on CFG Distillation strategy and Semantic Alignment in TwinFlow

### ISSUE THREAD

**Issue by ChuxiJ** [inclusionAI/TwinFlow#8]
Hi authors,

Congratulations on the impressive work! I am currently trying to reproduce TwinFlow and apply it to a pre-trained Flow Matching model.

I have a critical question regarding how Classifier-Free Guidance (CFG) is handled during the training process, which seems to be missing from the paper.

State-of-the-art distillation methods (e.g., DMD2, LCM, InstaFlow) typically "bake" the CFG effect into the student model. This is usually done by using the teacher's CFG-guided velocity.

Without this, 1-step generations from standard FM/Diffusion models usually suffer from poor semantic alignment and weak visual impact.

Is CFG applied when calculating the training targets?

Thanks!

**Comment by QAQdev** [comment]
@ChuxiJ 
We have provided a tutorial on MNIST to help better understand the core implementation, you can find it here: https://github.com/inclusionAI/TwinFlow?tab=readme-ov-file#tutorials, please also see https://github.com/inclusionAI/TwinFlow?tab=readme-ov-file#-open-source-plans .

**Comment by ChuxiJ** [comment]
@QAQdev 

<img width="3134" height="814" alt="Image" src="https://github.com/user-attachments/assets/0ecfa7b3-892f-4690-817b-fd47f4959165" />

I am currently conducting a reproduction experiment using your latest code. It is my contention that your approach to CFG processing will result in variance explosion. From a theoretical perspective: your training model is consistently fitting the CFG-augmented target smoothed by EMA. Additionally, the EMA weight, with a decay rate of 0.99, undergoes iterative self-updates. After approximately several hundred training steps, it can be observed that the conditional prediction value (v) of the EMA itself already incorporates the CFG target. Reapplying the CFG formula for further enhancement will induce over-saturation and an excessive expansion of variance. Following a few training cycles, the entire task will fail to converge (collapse). This constitutes a critical issue—the EMA update process is continuously intensifying the strength of its inherent CFG effect.

**Comment by ChuxiJ** [comment]
I’m not entirely sure if continuing the training will bring the variance down—even with DMD loss and self-adversarial training. But here’s my concern: if the model keeps making wrong predictions and collapsing, is self-adversarial training even meaningful? Have you guys run into this issue before? Should I align my implementation with yours, or just keep training it?

**Comment by ChuxiJ** [comment]
Reaching out privately for some help. If this works, it’ll be the first open-source music generation model built with your method. 

Here’s our Discord link—you can test out our Text2Music model that uses the DMD2 algorithm. It currently runs on 8-step inference. I want to use your method to convert it into a low-step generative flow model. As you know, the DMD2 approach transforms continuous flow models into discrete ones, which creates a lot of limitations.

https://discord.gg/DD6FScKk

**Comment by ChuxiJ** [comment]
> mentioned.

That's absolutely fantastic! Feel free to DM me directly on Discord—I'm junmingong. You can try out our Discord bot on the playground.
